File: backend/src/main/java/com/backend/petplace/domain/place/service/PlaceService.java
package com.backend.petplace.domain.place.service;

import com.backend.petplace.domain.place.dto.response.PlaceDetailResponse;
import com.backend.petplace.domain.place.dto.response.PlaceSearchResponse;
import com.backend.petplace.domain.place.entity.Category2Type;
import com.backend.petplace.domain.place.entity.Place;
import com.backend.petplace.domain.place.projection.PlaceSearchRow;
import com.backend.petplace.domain.place.repository.PlaceRepository;
import com.backend.petplace.global.exception.BusinessException;
import com.backend.petplace.global.response.ErrorCode;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class PlaceService {

  private final PlaceRepository placeRepository;

  private static final int DEFAULT_RADIUS_KM = 10;
  private static final int MAX_RADIUS_KM = 30;
  private static final int DEFAULT_SIZE = 300;

  @Transactional(readOnly = true)
  public List<PlaceSearchResponse> searchPlaces(
      double lat, double lon,
      Integer radiusKm,
      List<Category2Type> category2List,
      String keyword
  ) {
    int rk = (radiusKm == null ? DEFAULT_RADIUS_KM : Math.min(radiusKm, MAX_RADIUS_KM));
    int radiusMeters = rk * 1_000;

    // 1) 바운딩 박스(BBox) 계산
    BBox box = bbox(lat, lon, rk);

    // 2) 카테고리2 문자열 목록
    List<String> cat2 = (category2List == null) ? List.of()
        : category2List.stream().map(Enum::name).toList();

    // 키워드 가공
    String normalizedKeyword = (keyword == null || keyword.isBlank()) ? null : keyword.trim();

    // 3) 쿼리 호출
    List<PlaceSearchRow> rows = placeRepository.searchWithinRadius(
        lat, lon, box.minLat(), box.maxLat(), box.minLon(), box.maxLon(), radiusMeters, cat2, cat2.size(),
        normalizedKeyword, DEFAULT_SIZE, 0);

    return rows.stream()
        .map(PlaceSearchResponse::from)
        .toList();
  }

  @Transactional(readOnly = true)
  public PlaceDetailResponse getPlaceDetail(Long placeId) {
    Place place = placeRepository.findById(placeId)
        .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND_PLACE));

    return PlaceDetailResponse.from(place);
  }

  private record BBox(double minLat, double maxLat, double minLon, double maxLon) {}

  // 바운딩박스 계산
  private static BBox bbox(double lat, double lon, double radiusKm) {
    double R = 6371.0; // km
    double dLat = Math.toDegrees(radiusKm / R);
    double dLon = Math.toDegrees(radiusKm / (R * Math.cos(Math.toRadians(lat))));
    return new BBox(lat - dLat, lat + dLat, lon - dLon, lon + dLon);
  }
}

File: backend/src/main/java/com/backend/petplace/domain/place/service/PlaceService.kt
package com.backend.petplace.domain.place.service

import com.backend.petplace.domain.place.dto.response.PlaceDetailResponse
import com.backend.petplace.domain.place.dto.response.PlaceSearchResponse
import com.backend.petplace.domain.place.entity.Category2Type
import com.backend.petplace.domain.place.repository.PlaceRepository
import com.backend.petplace.global.exception.BusinessException
import com.backend.petplace.global.response.ErrorCode
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class PlaceService(
    private val placeRepository: PlaceRepository
) {

    @Transactional(readOnly = true)
    fun searchPlaces(
        lat: Double,
        lon: Double,
        radiusKm: Int?,
        category2List: List<Category2Type>?,
        keyword: String?
    ): List<PlaceSearchResponse> {
        // radiusKm가 null이면 기본값, 최대값은 30km로 제한
        val rk = (radiusKm ?: DEFAULT_RADIUS_KM).coerceAtMost(MAX_RADIUS_KM)
        val radiusMeters = rk * 1_000

        // 1) 바운딩 박스 계산
        val box = bbox(lat, lon, rk.toDouble())

        // 2) 카테고리2 문자열 목록 (enum.name 사용)
        val cat2: List<String> = category2List
            ?.map { it.name }
            ?: emptyList()

        // 3) 키워드 가공 (null 또는 blank면 null로 통일)
        val normalizedKeyword = keyword
            ?.takeIf { it.isNotBlank() }
            ?.trim()

        // 4) 쿼리 호출
        val rows = placeRepository.searchWithinRadius(
            lat = lat,
            lon = lon,
            minLat = box.minLat,
            maxLat = box.maxLat,
            minLon = box.minLon,
            maxLon = box.maxLon,
            radiusMeters = radiusMeters,
            category2List = cat2,
            category2Count = cat2.size,
            keyword = normalizedKeyword,
            limit = DEFAULT_SIZE,
            offset = 0
        )

        return rows.map { PlaceSearchResponse.from(it) }
    }

    @Transactional(readOnly = true)
    fun getPlaceDetail(placeId: Long): PlaceDetailResponse {
        val place = placeRepository.findById(placeId)
            .orElseThrow { BusinessException(ErrorCode.NOT_FOUND_PLACE) }

        return PlaceDetailResponse.from(place)
    }

    // Kotlin 내부 data class로 BBox 정의
    private data class BBox(
        val minLat: Double,
        val maxLat: Double,
        val minLon: Double,
        val maxLon: Double
    )

    // 바운딩 박스 계산
    private fun bbox(lat: Double, lon: Double, radiusKm: Double): BBox {
        val R = 6371.0 // km
        val dLat = Math.toDegrees(radiusKm / R)
        val dLon = Math.toDegrees(radiusKm / (R * Math.cos(Math.toRadians(lat))))
        return BBox(
            minLat = lat - dLat,
            maxLat = lat + dLat,
            minLon = lon - dLon,
            maxLon = lon + dLon
        )
    }

    companion object {
        private const val DEFAULT_RADIUS_KM = 10
        private const val MAX_RADIUS_KM = 30
        private const val DEFAULT_SIZE = 300
    }
}
